### ISSUE TITLE
`clip-retrieval filter` webdataset support

### ISSUE THREAD

**Issue by afiaka87** [rom1504/clip-retrieval#27]
Having trouble debugging this; but I think after looking at the code briefly - there's not any webdataset retrieval I dont think?

The root problem is that I am not getting any images saved from the query; even though it finds some in the output.

**Comment by rom1504** [comment]
yeah indeed, see https://github.com/rom1504/clip-retrieval/issues/2
the current implementation of clip-retrieval filter does not scale to large datasets

there are several ways to implement it with different trade offs that may make sense depending on the use case:
1. precompute the results with the index, then go through the whole webdataset collection to extract a subwebdataset (useful if your result set is in similar size to the webdataset collection)
2. compute the results with the index, then save the urls, and redownload them with img2dataset : useful if your result set is significantly smaller than the webdataset collection
3. compute the results with the index, then get the metadata out of the parquet file or the hdf5 file (currently computed in clip back) : useful if you only want the metadata

I intend to play with the various solutions to make it convenient and efficient to extract subset, but if you want to try your hand at it, that would be welcome!

**Comment by afiaka87** [comment]
@rom1504 Haven't had a chance to look into this but do let me know how your experiments go as I think it's a great solution for people who need data in-advance (TPU's/large batch size)

**Comment by rom1504** [comment]
yeah I definitely intend to get this happening. I won't have time for now, but I'll probably get it done in a week or something

**Comment by afiaka87** [comment]
The "zpadded_key" would just the laion400m "image_path" - but I wasn't sure that was uniform and I noticed the metadata also has an equivalent "key" field. In case that's confusing.

**Comment by rom1504** [comment]
Interesting!
Filtering the tars is going to be pretty slow but that's indeed one possibility
I've been working on #31 recently, I will probably include filtering options into that too

**Comment by rom1504** [comment]
For datasets of the kind of laion (very large) it doesn't matter a lot of we lose some % of the data, that's why redownloading subsets seem to make sense to me.
Filtering the tars require having 10TB of space available, so it doesn't seem like the best strategy in addition to the speed concern.

That said, until I have some actual code to propose, it's a good thing you have a working first solution :)

I hope to be able to get this done soon but I'm a bit busy these days, hard to find the time.

**Comment by rom1504** [comment]
still intend to make batch metadata retrieval much faster so this is more viable

**Comment by xilongzhou** [comment]
Hi, I check the [simple_filter.ipynb file](https://github.com/rom1504/clip-retrieval/blob/main/notebook/simple_filter.ipynb), my question is that where can we download this `/home/rom1504/indices_paths.json`? Is this the index file of Laion5B or Laion400M? Thank you very much!

**Comment by csiki** [comment]
Here's a version of `clip_filter` I just implemented, that handles webdataset files locally or on s3, as I described above.
[https://gist.github.com/csiki/1925f6a16a7000a3e47cb320fbb90c52](https://gist.github.com/csiki/1925f6a16a7000a3e47cb320fbb90c52)

As @rom1504 said, it's still a little slow, especially when filtering s3 datasets. I'll implement a multiprocessing version soon to alleviate that somewhat.
